drivers: adc: rework Nordic SAADC driver

jaz1-nordic opened the pull request:
SHIM Nordic modification for the ADC driver controlling the SAADC peripheral. Replaced HAL based implementation in favor of nrfx driver.

As a next step, it is planned to implement a feature that will allow the peripheral SAADC timer to be used for sampling, and for this it is necessary to use the nrfx driver in this SHIM.
This will allow more accurate and faster sampling than the kernel mechanism currently provides.

Based on #41721 - rebased on main, comments resolved, updated.

nika-nordic commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -521,99 +590,45 @@ static int start_read(const struct device *dev,
-	nrf_saadc_buffer_init(NRF_SAADC,
-			      (nrf_saadc_value_t *)sequence->buffer,
-			      active_channels);
+	/* Function needs buffer pointer and number of samples not buffer
```
this comment does not exactly corresponds to what the function below does, so it might be confusing (it is at least for me).

What about:
```
	/* Buffer is filled in chunks, each chunk composed of number of samples
         * equal to number of active channels. Buffer pointer is advanced and reloaded after each chunk.
         */

```

nika-nordic commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -197,281 +204,342 @@ static int saadc_pm_hook(const struct device *dev, enum pm_device_action action)
 	if (!repeat) {
 #if defined(ADC_BUFFER_IN_RAM)
 		m_data.user_buffer = (uint8_t *)m_data.user_buffer +
 			samples_to_bytes(&ctx->sequence, nrfy_saadc_amount_get(NRF_SAADC));
```
can't we get rid of this `nrfy_saadc_amount_get` given that we have access to `m_data.event` ?

jaz1-nordic commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -6,18 +6,18 @@
+LOG_MODULE_REGISTER(adc_nrfx_saadc, CONFIG_ADC_LOG_LEVEL);
 
 #define DT_DRV_COMPAT nordic_nrf_saadc
 
```
Will be done in https://nordicsemi.atlassian.net/browse/NRFX-7708

jaz1-nordic commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -197,281 +204,342 @@ static int saadc_pm_hook(const struct device *dev, enum pm_device_action action)
-	};
-	uint8_t channel_id = channel_cfg->channel_id;
-	uint32_t input_negative = channel_cfg->input_negative;
+#endif
```
Will be done in https://nordicsemi.atlassian.net/browse/NRFX-7708

jaz1-nordic commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -197,281 +204,342 @@ static int saadc_pm_hook(const struct device *dev, enum pm_device_action action)
+		if (analog_input > SAADC_CH_PSELP_PSELP_VDDHDIV5) {
+#else
+		if (analog_input > SAADC_CH_PSELP_PSELP_VDD) {
+#endif
```
Will be done in https://nordicsemi.atlassian.net/browse/NRFX-7708

anangl commented:
> @anangl please look here first: #41721.

What am I supposed to see there?

jaz1-nordic committed:
drivers: adc: rework Nordic SAADC driver

SHIM Nordic modification for the ADC driver controlling the SAADC
peripheral. Replaced HAL based implementation in favor of nrfx driver.

As a next step, it is planned to implement a feature that will allow
the peripheral SAADC timer to be used for sampling, and for this it is
necessary to use the nrfx driver in this SHIM.
This will allow more accurate and faster sampling than the kernel
mechanism currently provides.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

jaz1-nordic commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -540,99 +570,45 @@ static int start_read(const struct device *dev,
-					    channel_id);
-				return -EINVAL;
-			}
-			/* Signal an error if the channel is configured as
-			 * single ended with a resolution which is identical
-			 * to the sample bit size. The SAADC's "single ended"
-			 * mode is really differential mode with the
-			 * negative input tied to ground. We can therefore
-			 * observe negative values if the positive input falls
-			 * below ground. If the sample bitsize is larger than
-			 * the resolution, we can detect negative values and
-			 * correct them to 0 after the sequencen has ended.
-			 */
-			if ((m_data.single_ended_channels & BIT(channel_id)) &&
-			    (NRF_SAADC_8BIT_SAMPLE_WIDTH == 8 && resolution == 8)) {
-				LOG_ERR("Channel %u invalid single ended resolution",
-					channel_id);
-				return -EINVAL;
-			}
```
This is unnecessary, as even for 8-bit mode the samples in memory are placed in 16-bit format.

anangl replied:
All changes related to special handling of 8-bit samples should be done in a separate commit with a proper explanation in the commit message.

jaz1-nordic commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -540,99 +570,45 @@ static int start_read(const struct device *dev,
-				NRF_SAADC_INPUT_DISABLED);
-		}
-	} while (++channel_id < SAADC_CH_NUM);
+	channels_cnt = POPCOUNT(nrfx_saadc_channels_configured_get() & selected_channels);
```
I have restored checking of configured channels.

jaz1-nordic committed:
samples: drivers: adc: Add variant for 8-bit resolution on nRF54x

Add sample variant to run code in 8-bit resolution on nRF54x series.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -669,58 +650,50 @@ static int adc_nrfx_read_async(const struct device *dev,
-#else
-		nrf_saadc_disable(NRF_SAADC);
-#endif
+	m_data.event = event;
```
This pointer should not be stored. Instead, `event->data.done.p_buffer` should be supplied to `correct_single_ended()` as a parameter and only `event->data.done.size` should be stored for later use in `adc_context_update_buffer_pointer()`.

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -182,315 +175,351 @@ static int adc_convert_acq_time(uint16_t acquisition_time, nrf_saadc_acqtime_t *
 	size_t needed_buffer_size;
 
-	needed_buffer_size = samples_to_bytes(sequence, active_channels);
+	needed_buffer_size = samples_to_bytes(sequence, channels_cnt);
```
`active_channel_cnt` should be used then to prevent that suggestion of being a bit mask; `channels_cnt` is too ambiguous, and not quite correct.

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -182,315 +168,351 @@ static int adc_convert_acq_time(uint16_t acquisition_time, nrf_saadc_acqtime_t *
+	if (err != 0) {
+		LOG_ERR("Cannot set acquisition time: %d", err);
 		return -EINVAL;
 	}
```
Are such double error messages really useful?

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -182,315 +168,351 @@ static int adc_convert_acq_time(uint16_t acquisition_time, nrf_saadc_acqtime_t *
+	if (analog_input > range_allowed) {
+		LOG_ERR("Invalid analog positive input number: %d", analog_input);
+		return -EINVAL;
+	} else if (analog_input < NRF_SAADC_AIN0) {
```
My comment was about both checks, not only `< NRF_SAADC_AIN0`. `range_allowed` is not quite right in the `SAADC_CH_PSELP_PSELP_VDDHDIV5` case, so I don't think there is a value in doing such checks when not necessary.

jaz1-nordic committed:
modules: hal_nordic: nrfx: Update NRFX API version to 3.12

Update NRFX_CONFIG_API_VER_MINOR to 12, to be compatible with
latest SAADC driver changes.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -673,58 +629,48 @@ static int adc_nrfx_read_async(const struct device *dev,
-#else
-		nrf_saadc_disable(NRF_SAADC);
-#endif
+	m_data.buffer = event->data.done.p_buffer;
```
nit: this should be actually done inside `if (event->type == NRFX_SAADC_EVT_DONE)`
